File: android/app/src/main/kotlin/com/cruisemesh/app/mesh/LanSubnetScan.kt
package com.cruisemesh.app.mesh

import java.net.Inet4Address
import java.net.InetAddress

// A cruise-ship LAN is commonly a single flat /16 (or bigger), so the
// user-initiated "Search my local network" action follows the network's own
// advertised prefix -- but never broader than a /16 (65,534 hosts), which
// bounds it to something a phone can finish in the background. A home /24
// still scans just its 254 hosts; only genuinely huge networks (e.g. a /8)
// get clamped down to a /16 around this phone's address. This ceiling is for
// the manual button only, explicitly allowed to be big since the user asked.
internal const val BROADEST_SCAN_PREFIX_LENGTH = 16

// Ceiling for the *automatic* full-subnet fallback sweep: a /16 is up to
// ~65k TCP probes, minutes of sustained radio at concurrency 64. A /20 is
// ~4,094 hosts, a much smaller unattended background cost. Ship/hotel Wi-Fi
// is exactly where the underlying network is a huge flat subnet, so the
// automatic path stays narrower than the manual one.
internal const val BROADEST_AUTOMATIC_SCAN_PREFIX_LENGTH = 20

// /31 and /32 have no usable host addresses and /30 already yields only two;
// clamp here so [subnetHosts] always has at least a couple of candidates.
internal const val NARROWEST_SCAN_PREFIX_LENGTH = 30

// Used when the network doesn't report a prefix length for our address.
internal const val DEFAULT_SCAN_PREFIX_LENGTH = 24

/** The prefix length [subnetHosts] actually scans for a network-reported [prefixLength]. */
internal fun effectiveScanPrefixLength(prefixLength: Int): Int =
    prefixLength.coerceIn(BROADEST_SCAN_PREFIX_LENGTH, NARROWEST_SCAN_PREFIX_LENGTH)

/**
 * Same clamp as [effectiveScanPrefixLength], but for the automatic
 * full-subnet sweep, which is capped narrower (see
 * [BROADEST_AUTOMATIC_SCAN_PREFIX_LENGTH]).
 */
internal fun effectiveAutomaticScanPrefixLength(prefixLength: Int): Int =
    prefixLength.coerceIn(BROADEST_AUTOMATIC_SCAN_PREFIX_LENGTH, NARROWEST_SCAN_PREFIX_LENGTH)

/**
 * Candidate hosts in the IPv4 subnet [localAddress] sits in, as defined by
 * [prefixLength] (clamped by [effectiveScanPrefixLength]), excluding this
 * phone plus the network and broadcast addresses.
 */
internal fun subnetHosts(localAddress: Inet4Address, prefixLength: Int): List<InetAddress> {
    val effectivePrefix = effectiveScanPrefixLength(prefixLength)
    val localValue = localAddress.address.toIpv4Long()
    // effectivePrefix is 16..30, so the shift is always in range and the mask
    // never spans the full 32 bits (which would need special-casing).
    val mask = (0xFFFF_FFFFL shl (32 - effectivePrefix)) and 0xFFFF_FFFFL
    val network = localValue and mask
    val broadcast = network or (mask.inv() and 0xFFFF_FFFFL)
    val hosts = ArrayList<InetAddress>(((broadcast - network - 1).coerceAtLeast(0)).toInt())
    var host = network + 1
    while (host < broadcast) {
        if (host != localValue) {
            hosts.add(InetAddress.getByAddress(host.toIpv4Bytes()))
        }
        host++
    }
    return hosts
}

private fun ByteArray.toIpv4Long(): Long =
    fold(0L) { acc, byte -> (acc shl 8) or (byte.toLong() and 0xFF) }

private fun Long.toIpv4Bytes(): ByteArray = byteArrayOf(
    ((this shr 24) and 0xFF).toByte(),
    ((this shr 16) and 0xFF).toByte(),
    ((this shr 8) and 0xFF).toByte(),
    (this and 0xFF).toByte(),
)

File: ios/CruiseMesh/Mesh/LanSubnetScan.swift
import Foundation

/// Ceiling breadth for the user-initiated "Search local subnet" action --
/// explicitly allowed to be big since the user asked for it.
let broadestLanScanPrefixLength = 16
/// Ceiling breadth for the *automatic* full-subnet fallback sweep. A /16 is
/// ~65k hosts, minutes of sustained radio at concurrency 64; a /20 is
/// ~4,094 hosts, a much smaller background cost. Ship/hotel Wi-Fi is
/// exactly where the underlying network is a huge flat subnet, so the
/// unattended path stays narrower than the manual one.
let broadestAutomaticLanScanPrefixLength = 20
let narrowestLanScanPrefixLength = 30
let defaultLanScanPrefixLength = 24

struct LocalWifiIPv4Network: Equatable {
    let address: String
    let prefixLength: Int
}

func effectiveLanScanPrefixLength(_ prefixLength: Int) -> Int {
    min(max(prefixLength, broadestLanScanPrefixLength), narrowestLanScanPrefixLength)
}

/// Same clamp as `effectiveLanScanPrefixLength`, but for the automatic
/// full-subnet sweep, which is capped narrower (see
/// `broadestAutomaticLanScanPrefixLength`).
func effectiveAutomaticLanScanPrefixLength(_ prefixLength: Int) -> Int {
    min(max(prefixLength, broadestAutomaticLanScanPrefixLength), narrowestLanScanPrefixLength)
}

/// Candidate hosts in the selected IPv4 subnet, excluding this phone plus the
/// network and broadcast addresses. Breadth is always clamped to /16.../30.
func lanSubnetHosts(localAddress: String, prefixLength: Int) -> [String] {
    guard let local = ipv4Value(localAddress) else { return [] }
    let effectivePrefix = effectiveLanScanPrefixLength(prefixLength)
    let mask = UInt32.max << UInt32(32 - effectivePrefix)
    let network = local & mask
    let broadcast = network | ~mask
    guard network < broadcast else { return [] }

    var hosts: [String] = []
    hosts.reserveCapacity(max(Int(broadcast - network - 2), 0))
    var host = network + 1
    while host < broadcast {
        if host != local {
            hosts.append(ipv4String(host))
        }
        host += 1
    }
    return hosts
}

func ipv4PrefixLength(netmask: UInt32) -> Int? {
    let inverse = ~netmask
    guard inverse & (inverse &+ 1) == 0 else { return nil }
    return netmask.nonzeroBitCount
}

private func ipv4Value(_ address: String) -> UInt32? {
    let parts = address.split(separator: ".", omittingEmptySubsequences: false)
    guard parts.count == 4 else { return nil }
    var value: UInt32 = 0
    for part in parts {
        guard let octet = UInt8(part) else { return nil }
        value = (value << 8) | UInt32(octet)
    }
    return value
}

private func ipv4String(_ value: UInt32) -> String {
    [24, 16, 8, 0]
        .map { String((value >> UInt32($0)) & 0xff) }
        .joined(separator: ".")
}
